### PR TITLE
Add support for constraint failures

### PR DESCRIPTION
**Pull request by pnwpedro**
## Problem

We don't support constraint_failures from the QueryFailure wire protocol.

## Solution

Error handling probably deserves more love soon, but I think this is an improvment.

* Add support for constraint_failures. For now, added ConstraintFailure dataclass to errors.py file, although the wire_protocol.py should be renamed to encoding.py and we should probably encapsulate wire wire protocol concepts in a wire protocol file..
* Separates protocol error handling from the rest for the moment and be strict with query success and failure
* Sneak in a quick stats contract update
* Add a few unit tests
* Add an integ test for the constraint failures

## Testing

Added new unit tests and a new integ test

### CODE REVIEW

File: fauna/client.py [hunk 1]
```diff
@@ -256,85 +259,118 @@ def _query(
...
-                    err.code,
-                    err.message,
-                    err.summary,
+                    status_code,
+                    code,
+                    message,
+                    summary,
+                )
+            else:
+                raise QueryRuntimeError(
```
Comment (cleve-fauna) [inline fauna/client.py:321]: out of scope for this PR.

But was talking with Paul about whether we should get rid of the QueryRuntime / QueryCheck distinction in the drivers.

There really the same semantically for the user - my query is no good.

Would also decouple us from having to map codes to error types.

Putting here for discussion.

File: fauna/errors.py [hunk 2]
```diff
@@ -116,6 +136,13 @@ class ServiceInternalError(ServiceError):
 
 
 class ServiceTimeoutError(ServiceError):
```
Comment (cleve-fauna) [inline fauna/errors.py:138]: should these be named `Exception` not `Error` to be more pythonic?

File: fauna/response.py [hunk 3]
```diff
@@ -50,44 +53,12 @@ def __init__(
...
         self._traceparent = headers.get("traceparent", "")
         self._headers = headers
-
-        if "data" not in response_json:
-            raise ProtocolError(
-                self._status_code,
-                "Unexpected response",
-                f"Key 'data' not found in response body: \n{response_json}",
-            )
+        self._stats: Mapping[Union[str, Stat], Any] = {}
```
Comment (cleve-fauna) [inline fauna/response.py:56]: can we assume the stats are ints?

**Review by cleve-fauna — APPROVED** [review]
some comments but non-blocking

File: fauna/errors.py (hunk 2 above)
Comment (pnwpedro) [reply fauna/errors.py:138]: possibly. i can revisit this

File: fauna/response.py (hunk 3 above)
Comment (pnwpedro) [reply fauna/response.py:56]: each stat has an individual data type declared, so this felt safer to me.

File: fauna/client.py (hunk 1 above)
Comment (pnwpedro) [reply fauna/client.py:321]: Yeah, I'm down for consolidating this to QueryError. If we see users asking for special types, we can always inherit from QueryError at that point

**Pull request merged**